File: smart-office/services/4-notification-service/src/sqs.ts
import { SQSClient, ReceiveMessageCommand, DeleteMessageCommand, Message } from '@aws-sdk/client-sqs';
import dotenv from 'dotenv';

dotenv.config();

const sqsClient = new SQSClient({ 
    region: process.env.AWS_REGION,
    // endpoint is needed for localstack or similar, remove for AWS
});
const queueUrl = process.env.SQS_QUEUE_URL;

const handleMessage = (message: Message) => {
    console.log('--- New Notification ---');
    if (message.Body) {
        try {
            const body = JSON.parse(message.Body);
            const event = JSON.parse(body.Message); // Message from SNS is a stringified JSON
            const booking = event.payload;
            
            console.log(`Received booking confirmation for User: ${booking.userId}`);
            console.log(`Resource Type: ${booking.bookingType}`);
            console.log(`Resource ID: ${booking.resourceId}`);
            console.log('------------------------');

        } catch (e) {
            console.error("Failed to parse message body:", e);
            console.log("Raw message body:", message.Body);
        }
    }
};

export const pollSqsQueue = async () => {
    if (!queueUrl) {
        console.error('SQS_QUEUE_URL is not defined. Stopping poller.');
        return;
    }

    while (true) {
        try {
            const { Messages } = await sqsClient.send(new ReceiveMessageCommand({
                QueueUrl: queueUrl,
                MaxNumberOfMessages: 10,
                WaitTimeSeconds: 20, // Long polling
            }));

            if (Messages && Messages.length > 0) {
                for (const message of Messages) {
                    handleMessage(message);
                    // Delete the message from the queue
                    await sqsClient.send(new DeleteMessageCommand({
                        QueueUrl: queueUrl,
                        ReceiptHandle: message.ReceiptHandle,
                    }));
                }
            }
        } catch (error) {
            console.error('Error polling SQS:', error);
            // Wait before retrying
            await new Promise(resolve => setTimeout(resolve, 5000));
        }
    }
};

File: smart-office/services/5-analytics-service/app/sqs.py
import asyncio
import boto3
import os
import json
from . import database
from datetime import datetime

AWS_REGION = os.getenv("AWS_REGION")
SQS_QUEUE_URL = os.getenv("SQS_QUEUE_URL")

sqs_client = boto3.client("sqs", region_name=AWS_REGION)

def save_booking_event(event_payload: dict):
    """Saves a booking event to the DynamoDB table."""
    try:
        database.table.put_item(
            Item={
                'booking_id': event_payload.get("bookingId"),
                'user_id': event_payload.get("userId"),
                'resource_id': event_payload.get("resourceId"),
                'booking_type': event_payload.get("bookingType"),
                'created_at': event_payload.get("createdAt")
            }
        )
        print(f"Successfully saved booking event {event_payload.get('bookingId')} to DynamoDB.")
    except Exception as e:
        print(f"Error saving to DynamoDB: {e}")

async def poll_sqs_queue():
    """Polls the SQS queue for messages and processes them."""
    if not SQS_QUEUE_URL:
        print("SQS_QUEUE_URL not set. SQS poller is disabled.")
        return

    print("Starting SQS poller for analytics...")

    while True:
        try:
            response = sqs_client.receive_message(
                QueueUrl=SQS_QUEUE_URL,
                MaxNumberOfMessages=10,
                WaitTimeSeconds=20
            )

            messages = response.get("Messages", [])
            if messages:
                for msg in messages:
                    try:
                        # SNS messages are nested inside the SQS message body
                        sns_msg = json.loads(msg["Body"])
                        event_payload = json.loads(sns_msg["Message"])["payload"]
                        
                        save_booking_event(event_payload)

                        # Delete message from queue
                        sqs_client.delete_message(
                            QueueUrl=SQS_QUEUE_URL,
                            ReceiptHandle=msg["ReceiptHandle"]
                        )
                    except Exception as e:
                        print(f"Error processing message: {e}")

        except Exception as e:
            print(f"Error polling SQS queue: {e}")
        
        await asyncio.sleep(5) # Wait before the next poll
